app/src/main/java/co/runloop/influencer/fragment/ContactsFragment.java

```diff
@@ -54,21 +58,37 @@ public View onCreateView(@NonNull LayoutInflater inflater,
         recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
         recyclerView.setAdapter(adapter);
 
+        progressBar = root.findViewById(R.id.frag_contacts_list_progressbar);
+
         contactsViewModel = ViewModelProviders
                 .of(this)
                 .get(ContactsViewModel.class);
-        contactsViewModel.getContacts().observe(this, (@Nullable List<Contact> contacts) -> {
-            adapter.setContacts(contacts);
-            adapter.notifyDataSetChanged();
-        });
-        if (contactsIsAvailable()) {
-            contactsViewModel.onPermissionGranted();
-            if(contactsViewModel.getContacts().getValue() == null) {
-                contactsViewModel.loadAll();
+        contactsViewModel.getContacts().observe(this, (@Nullable Resource<List<Contact>> contacts) -> {
+            if (contacts == null) {
+                progressBar.setVisibility(View.GONE);
+                return;
             }
-        }
-        if (!contactsIsAvailable()) {
-            requestReadContactsPermission();
+            switch (contacts.getStatus()) {
+                case Progress:
+                    progressBar.setVisibility(View.VISIBLE);
+                    break;
+                case Completed:
+                    adapter.setContacts(contacts.getData());
+                    adapter.notifyDataSetChanged();
+                    progressBar.setVisibility(View.GONE);
+                    break;
+                case Error:
+                    if (contacts.getError() instanceof SecurityException && !contactsIsAvailable()) {
+                        requestReadContactsPermission();
+                    }
+                    progressBar.setVisibility(View.GONE);
+                    break;
+            }
+        });
+
+        if (contactsViewModel.getContacts().getValue() == null
```

Comment:
Do you think it makes sense to wrap this condition in some `areContactsAvailable` method in the view model?

Reply:
ok

Reply:
Well, I was just thinking out loud. It's a question, not an instruction :) 
But if you agree...

Reply:
I just think that is right )

Reply:
Cool 😄